README.md
# Phantom - Client

<p align="center">
    <img src="./banner.png?sanitize=true" />
</p>

## Installation

### Node.js

```bash
yarn add @phantomchain/client
```

If you want to use the CDN version:

```html
<script src="https://unpkg.com/@phantomchain/client/dist/bundle.umd.js"></script>
```

## Usage

Import the library in Node.js:

```
import PhantomChainClient from @phantomchain/client
```
or
```
const PhantomChainClient require('@phantomchain/client')
```

Use the library:

```
const client = new PhantomChainClient('<your host here>')
```

### Examples
There are more examples in the `examples` folder.

## Security

If you discover a security vulnerability within this package, please send an e-mail to security@ark.io. All security vulnerabilities will be promptly addressed.

## Credits

- [Brian Faust](https://github.com/faustbrian)
- [Lúcio Rubens](https://github.com/luciorubeens)
- [Alex Barnsley](https://github.com/alexbarnsley)
- [Juan A. Martín](https://github.com/j-a-m-l)
- [All Contributors](../../../../contributors)

## License

[MIT](LICENSE) © [PhantomChain](https://phantom.org)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

## 0.1.20 - 2019-01-28

### Fixed
- Check status and version to find peers
- Clone peer list as to not overwrite default network peers

## 0.1.19 - 2019-01-25

### Changed
- Added transaction fees v2 endpoint

## 0.1.18 - 2019-01-22

### Changed
- Allow https peer option when finding peers

## 0.1.17 - 2019-01-16

### Changed
- Use standard API-Version header

## 0.1.16 - 2018-12-04

### Fixed
- Update mainnet peer addresses to match Core

## 0.1.15 - 2018-12-04

### Fixed
- Use API port for mainnet peers

## 0.1.14 - 2018-11-27

### Changed
- Update mainnet peer list

## 0.1.13 - 2018-11-26

### Changed
- Update mainnet peer list

## 0.1.12 - 2018-11-23

### Changed
- Get peer list from 2 peers

## 0.1.11 - 2018-11-23

### Changed
- Always send the `Content-Type` header

## 0.1.10 - 2018-10-24

### Fixed
- Use the configured `core-api` port instead of the `core-p2p` port, which would fail (https://github.com/PhantomChain/core/pull/1138)
- Wrong loader methods (https://github.com/PhantomChain/core/pull/1194)

## 0.1.9 - 2018-10-13

### Fixed
- Handle v2 endpoints when finding peers (https://github.com/PhantomChain/core/pull/1103)

## 0.1.8 - 2018-10-12

### Changed
- Use 5 second timeout for finding peers (https://github.com/PhantomChain/core/pull/1103)

### Fixed
- Get peer response data correctly (https://github.com/PhantomChain/core/pull/1103)

## 0.1.7 - 2018-10-12

### Fixed
- Add query parameters to the v2 endpoints.(https://github.com/PhantomChain/core/pull/1103)

## 0.1.6 - 2018-10-09

### Fixed
- Use the definitive `accept` header instead of the previous one (https://github.com/PhantomChain/core/pull/1082)

## 0.1.5 - 2018-10-05

### Fixed
- Use the `accept` header instead of api-version to avoid CORS problems (https://github.com/PhantomChain/core/pull/1012)

## 0.1.4 - 2018-09-20

### Fixed
- Fix API client HTTP params (query string) (https://github.com/PhantomChain/core/pull/1015)

## 0.1.3

...

## 0.1.1 - 2018-06-14

### Added
- initial release


__tests__/http.test.js
const axios = require('axios')
const MockAdapter = require('axios-mock-adapter')
const mock = new MockAdapter(axios)

const toHaveAtLeastHeaders = require('./matchers/http/headers')
expect.extend({ toHaveAtLeastHeaders })

const HttpClient = require('../lib/http')

const host = 'http://example.net'
let client

beforeEach(() => {
  client = new HttpClient(host, 2)
})

describe('API - HTTP Client', () => {
  let defaultHeaders

  beforeEach(() => {
    defaultHeaders = {
      'API-Version': client.version
    }
  })

  describe('constructor', () => {
    it('should be instantiated', () => {
      expect(client).toBeInstanceOf(HttpClient)
    })

    describe('host', () => {
      it('should set the host', () => {
        client = new HttpClient(host)
        expect(client.host).toBe(host)
      })

      it('should remove the final slash of the host when necessary', () => {
        client = new HttpClient('http://example.net/')
        expect(client.host).toBe('http://example.net')
      })

      it('should check that the host is not empty', () => {
        expect(() => new HttpClient('')).toThrow()
      })
    })

    describe('API version parameter', () => {
      it('should set the API version', () => {
        client = new HttpClient(host, 3)
        expect(client.version).toBe(3)
      })

      it('should use 2 when is not present', () => {
        client = new HttpClient(host)
        expect(client.version).toBe(2)
      })
    })
  })

  describe('setTimeout', () => {
    it('should set the timeout', () => {
      expect(client.timeout).toBe(60000)
      client.setTimeout(5000)
      expect(client.timeout).toBe(5000)
    })
  })

  describe('setHeaders', () => {
    it('should set the headers', () => {
      const newHeaders = {
        'API-Version': 30,
        other: 'value'
      }
      client.setHeaders(newHeaders)

      expect(client.headers).toEqual(newHeaders)
    })
  })

  describe('get', () => {
    beforeEach(() => {
      mock.onGet(`${host}/api/ENDPOINT`).reply(200, { data: [] })
    })

    it('should send GET requests to the API', async () => {
      const response = await client.get('ENDPOINT')

      expect(response.status).toBe(200)
    })

    it('should use the necessary request headers', async () => {
      const response = await client.get('ENDPOINT')

      expect(response.config).toHaveAtLeastHeaders(defaultHeaders)
    })

    it('should send the request params', async () => {
      const params = { param1: 'value1', param2: 'value2' }

      mock.reset()
      mock.onGet(`${host}/api/ENDPOINT`, { params }).reply(200, { data: [] })

      const response = await client.get('ENDPOINT', params)

      expect(response.status).toBe(200)
    })
  })

  describe('post', () => {
    beforeEach(() => {
      mock.onPost(`${host}/api/ENDPOINT`).reply(200, { data: [] })
    })

    it('should send POST requests to the API', async () => {
      const response = await client.post('ENDPOINT')

      expect(response.status).toBe(200)
    })

    it('should use the necessary request headers', async () => {
      const response = await client.post('ENDPOINT')

      expect(response.config).toHaveAtLeastHeaders(defaultHeaders)
    })
  })

  describe('put', () => {
    beforeEach(() => {
      mock.onPut(`${host}/api/ENDPOINT`).reply(200, { data: [] })
    })

    it('should send PUT requests to the API', async () => {
      const response = await client.put('ENDPOINT')

      expect(response.status).toBe(200)
    })

    it('should use the necessary request headers', async () => {
      const response = await client.put('ENDPOINT')

      expect(response.config).toHaveAtLeastHeaders(defaultHeaders)
    })
  })

  describe('patch', () => {
    beforeEach(() => {
      mock.onPatch(`${host}/api/ENDPOINT`).reply(200, { data: [] })
    })

    it('should send PATCH requests to the API', async () => {
      const response = await client.patch('ENDPOINT')

      expect(response.status).toBe(200)
    })

    it('should use the necessary request headers', async () => {
      const response = await client.patch('ENDPOINT')

      expect(response.config).toHaveAtLeastHeaders(defaultHeaders)
    })
  })

  describe('delete', () => {
    beforeEach(() => {
      mock.onDelete(`${host}/api/ENDPOINT`).reply(200, { data: [] })
    })

    it('should send DELETE requests to the API', async () => {
      const response = await client.delete('ENDPOINT')

      expect(response.status).toBe(200)
    })

    it('should use the necessary request headers', async () => {
      const response = await client.delete('ENDPOINT')

      expect(response.config).toHaveAtLeastHeaders(defaultHeaders)
    })

    it('should send the request params', async () => {
      const params = { param1: 'value1', param2: 'value2' }

      mock.reset()
      mock.onDelete(`${host}/api/ENDPOINT`, { params }).reply(200, { data: [] })

      const response = await client.delete('ENDPOINT', params)

      expect(response.status).toBe(200)
    })
  })
})


__tests__/client.test.js
const axios = require('axios')
const MockAdapter = require('axios-mock-adapter')
const httpMock = new MockAdapter(axios)
const { arrayContaining } = expect

const Client = require('../lib')
const HttpClient = require('../lib/http')
const ApiResource = require('../lib/resources/api/transactions')
const initialPeers = require('../lib/peers')

// https://github.com/facebook/jest/issues/3601
const errorCapturer = fn => fn.then(res => () => res).catch(err => () => { throw err })

const host = 'https://example.net:4003'
const { peers, peersOverride } = require('./fixtures/peers')

jest.mock('../lib/peers', () => {
  const { initialPeers } = require('./fixtures/peers')
  return { mainnet: initialPeers, devnet: initialPeers }
})

let client

beforeEach(() => {
  httpMock.reset()
  client = (new Client(host))
})

describe('API - Client', () => {
  describe('constructor', () => {
    it('should be instantiated', () => {
      expect(client).toBeInstanceOf(Client)
    })

    it('should set connection', () => {
      expect(client.http).toBeInstanceOf(HttpClient)
    })

    it('should return an API resource', () => {
      expect(client.resource('transactions')).toBeInstanceOf(ApiResource)
    })

    it('should use 1 as the default API version', () => {
      expect(client.version).toBe(2)
    })

    it('should set the API version', () => {
      client = (new Client('https://localhost:4003', 3))
      expect(client.version).toBe(3)
    })

    it('should set the HTTP client API version', () => {
      client = (new Client('https://localhost:4003', 3))
      expect(client.getConnection().version).toBe(3)
    })
  })

  describe('setVersion', () => {
    it('should set the API version', () => {
      client.setVersion(2)

      expect(client.version).toBe(2)
    })

    it('should set the API version of the HTTP client too', () => {
      client.setVersion(2)

      expect(client.http.version).toBe(2)
    })

    it('should throw an Error if the API version is falsy', async () => {
      expect(() => client.setVersion(0)).toThrowError(/api.*version/i)
    })
  })

  describe('static findPeers', () => {
    it('should throw an Error if the network does not have initial peers', async () => {
      expect(await errorCapturer(Client.findPeers('wrong'))).toThrowError(/network.*wrong/i)
    })

    xit('should connect randomly to the initial peers', () => {
    })

    describe('on version 2', () => {
      it('should return a sorted list of peers', async () => {
        const apiPeers = peers.map(peer => {
          return {
            ...peer,
            status: 200,
            version: '2.0.0'
          }
        })

        const data = {
          meta: {},
          data: apiPeers
        }

        httpMock.onGet(/http.*\/api\/peers/).reply(200, data)

        const foundPeers = await Client.findPeers('devnet', 2)

        expect(foundPeers).toEqual([
          apiPeers[1],
          apiPeers[3],
          apiPeers[0],
          apiPeers[2]
        ])
      })
    })

    it('should ignore local peers', async () => {
      const ips = ['127.0.0.1', '::ffff:127.0.0.1', '::1']

      expect.assertions(ips.length * 2)

      for (const ip of ips) {
        const localPeer = {
          ip,
          height: 3663605,
          status: 200,
          latency: 17
        }
        const data = {
          data: peers.concat([localPeer])
        }

        httpMock.onGet(/http.*\/api\/peers/).reply(200, data)

        const foundPeers = await Client.findPeers('devnet')
        expect(foundPeers).toEqual(arrayContaining(peers))
        expect(foundPeers).not.toContainEqual(localPeer)
      }
    })

    it('should ignore not-OK peers', async () => {
      const notOkPeer = {
        ip: '7.7.7.7',
        port: 3333,
        height: 3663605,
        status: 'not OK',
        delay: 17
      }
      const data = {
        data: peers.concat([notOkPeer])
      }

      httpMock.onGet(/http.*\/api\/peers/).reply(200, data)

      const foundPeers = await Client.findPeers('devnet')
      expect(foundPeers).toEqual(arrayContaining(peers))
      expect(foundPeers).not.toContainEqual(notOkPeer)
    })

    it('should check 2 peers', async () => {
      const data = {
        data: peers
      }

      const peerRespond = jest.fn(() => { return [200, data] })
      httpMock.onGet(/http.*\/api\/peers/).reply(() => (peerRespond()))

      const foundPeers = await Client.findPeers('devnet')
      expect(foundPeers).toEqual(arrayContaining(peers))
      expect(peerRespond).toHaveBeenCalledTimes(2)
    })

    it('should check https if specified', async () => {
      const data = {
        success: true,
        peers: peers
      }

      const httpPeer = peers[0]
      const httpsPeer = { ...peers[1], isHttps: true }

      const httpRespond = jest.fn(() => { return [200, data] })
      const httpsRespond = jest.fn(() => { return [200, data] })
      httpMock.onGet(`http://${httpPeer.ip}:${httpPeer.port}/api/peers`).reply(() => (httpRespond()))
      httpMock.onGet(`https://${httpsPeer.ip}:${httpsPeer.port}/api/peers`).reply(() => (httpsRespond()))

      await Client.findPeers(null, 2, [httpPeer, httpsPeer])
      expect(httpRespond).toHaveBeenCalledTimes(1)
      expect(httpsRespond).toHaveBeenCalledTimes(1)
    })

    xdescribe('when a peer is not valid', () => {
      it('tries others', () => {
      })
    })

    describe('when the request to find peers fails', () => {
      beforeEach(() => {
        httpMock.onGet(/http.*\/api\/peers/).reply(500)
      })

      it('returns the list of initial (hardcoded) peers', async () => {
        const foundPeers = await Client.findPeers('devnet')
        expect(foundPeers).not.toEqual(arrayContaining(peers))
        expect(foundPeers).toEqual(arrayContaining(initialPeers.devnet))
      })

      it('should return the list of provided peers', async () => {
        const foundPeers = await Client.findPeers('devnet', 2, peersOverride)
        expect(foundPeers).not.toEqual(arrayContaining(peers))
        expect(foundPeers).toEqual(arrayContaining(peersOverride))
      })
    })
  })

  describe('fetchPeerConfig', () => {
    const peer = peers[0]
    const host = `http://${peer.ip}:${peer.port}`

    it('should obtain the configuration of the peer', async () => {
      const peerConfig = { plugins: {} }
      httpMock.onGet(`${host}/config`).reply(200, { data: peerConfig })

      const config = await Client.fetchPeerConfig(host)
      expect(config).toEqual(peerConfig)
    })

    describe('when there is a network error', () => {
      it('should not throw Errors', async () => {
        httpMock.onGet(`${host}/config`).reply(500, {})

        expect(await errorCapturer(Client.fetchPeerConfig(host))).not.toThrowError()
      })
    })
  })

  describe('selectApiPeer', () => {
    describe('when there is at least 1 peer with `core-api` enabled', () => {
      it('should obtain the configuration of that peer', async () => {
        const peer = peers[1]
        const port = 10101
        const peerConfig = {
          plugins: {
            '@phantomchain/core-api': { enabled: true, port }
          }
        }
        httpMock.onGet(`http://${peer.ip}:${peer.port}/config`).reply(200, { data: peerConfig })
        httpMock.onGet(/http.*\/config/).reply(200, { data: { plugins: {} } })

        expect(await Client.selectApiPeer(peers)).toEqual({
          ...peer,
          port
        })
      })
    })

    describe('when there are not peers with the `core-api` plugin', () => {
      it('should return `null`', async () => {
        const peerConfig = {
          plugins: {
            '@phantomchain/core-other-plugin': { enabled: true }
          }
        }
        httpMock.onGet(/http.*\/config/).reply(200, { data: peerConfig })

        expect(await Client.selectApiPeer(peers)).toBeNull()
      })
    })

    describe('when there are not peers with `core-api` enabled', () => {
      it('should return `null`', async () => {
        const peerConfig = {
          plugins: {
            '@phantomchain/core-api': { enabled: false }
          }
        }
        httpMock.onGet(/http.*\/config/).reply(200, { data: peerConfig })

        expect(await Client.selectApiPeer(peers)).toBeNull()
      })
    })

    describe('when there is a network error', () => {
      it('should not throw Errors', async () => {
        httpMock.onGet(/http.*\/config/).reply(500, {})

        expect(await errorCapturer(Client.selectApiPeer(peers))).not.toThrowError()
      })
    })
  })

  describe('static connect', () => {
    const port = 10101
    const peerConfig = {
      plugins: {
        '@phantomchain/core-api': { enabled: true, port }
      }
    }

    it('should throw an Error if the network does not have initial peers', async () => {
      httpMock.onGet(/http.*\/config/).reply(200, { data: peerConfig })

      try {
        expect(async () => Client.connect('wrong')).toThrow()
        expect().fail('Should fail on the previous line')
      } catch (error) {
        expect(error).toBeInstanceOf(Error)
      }
    })

    describe('when the peers are overriden', () => {
      it('should not throw an Error if the network does not exist', async () => {
        httpMock.onGet(/http.*\/config/).reply(200, { data: peerConfig })

        try {
          await Client.connect('wrong', 2, peersOverride)
        } catch (error) {
          expect().fail('Should not have failed on the previous line')
        }
      })

      it('should not use the seed peers of the network', () => {
      })
    })

    describe('when there is at least 1 peer with `core-api` enabled', () => {
      it('should find peers and select the most updated and less delayed', async () => {
        const data = {
          data: peers
        }
        httpMock.onGet(/http.*\/config/).reply(200, { data: peerConfig })
        httpMock.onGet(/http.*\/api\/peers/).reply(200, data)

        const client = await Client.connect('devnet')
        expect(client.getConnection().host).toEqual(`http://${peers[1].ip}:${port}`)
      })

      it('should select a peer with `core-api` enabled', async () => {
        const peer = peers[2]
        const data = {
          data: peers
        }

        httpMock.onGet(`http://${peer.ip}:${peer.port}/config`).reply(200, { data: peerConfig })
        httpMock.onGet(/http.*\/config/).reply(200, { data: { plugins: {} } })
        httpMock.onGet(/http.*\/api\/peers/).reply(200, data)

        const client = await Client.connect('devnet')
        expect(client.getConnection().host).toEqual(`http://${peer.ip}:${port}`)
      })
    })

    describe('when there are not peers with `core-api` enabled', () => {
      it('should throw an Error', async () => {
        httpMock.onGet(/http.*\/config/).reply(200, { data: {} })

        expect(await errorCapturer(Client.connect('devnet', 2))).toThrowError(/api/i)
      })
    })
  })
})


__tests__/resources/transactions.test.js
const Client = require('../../lib')
const ApiResource = require('../../lib/resources/api/transactions')

const configureMocks = require('../mocks')
const host = 'https://example.net:4003'
configureMocks({ host })

let resource

beforeEach(() => {
  resource = (new Client(host)).setVersion(2).resource('transactions')
})

describe('API - 2.0 - Resources - Transactions', () => {
  it('should be instantiated', () => {
    expect(resource).toBeInstanceOf(ApiResource)
  })

  it('should call "all" method', async () => {
    const response = await resource.all()

    expect(response.status).toBe(200)
  })

  it('should call "create" method', async () => {
    const response = await resource.create()

    expect(response.status).toBe(200)
  })

  it('should call "get" method', async () => {
    const response = await resource.get('123')

    expect(response.status).toBe(200)
  })

  it('should call "allUnconfirmed" method', async () => {
    const response = await resource.allUnconfirmed()

    expect(response.status).toBe(200)
  })

  it('should call "getUnconfirmed" method', async () => {
    const response = await resource.getUnconfirmed('123')

    expect(response.status).toBe(200)
  })

  it('should call "search" method', async () => {
    const response = await resource.search({})

    expect(response.status).toBe(200)
  })

  it('should call "types" method', async () => {
    const response = await resource.types()

    expect(response.status).toBe(200)
  })

  it('should call "fees" method', async () => {
    const response = await resource.fees()

    expect(response.status).toBe(200)
  })
})


lib/http.js
const axios = require('axios')

module.exports = class HttpClient {
  /**
   * @constructor
   * @param  {String} host
   * @param  {Number = 2} [apiVersion]
   */
  constructor (host, apiVersion = 2) {
    this.host = host.endsWith('/') ? host.slice(0, -1) : host

    if (host.length === 0) {
      throw new Error('An empty host is not permitted')
    }

    this.version = apiVersion
    this.timeout = 60000
    this.headers = { 'Content-Type': 'application/json' }
  }

  /**
   * Used to specify the API Version.
   * @param {Number} version
   */
  setVersion (version) {
    this.version = version
  }

  /**
   * Establish the headers of the requests.
   * @param {Object} headers
   */
  setHeaders (headers = {}) {
    this.headers = headers
  }

  /**
   * Establish the timeout of the requests.
   * @param {Number} timeout
   */
  setTimeout (timeout) {
    this.timeout = timeout
  }

  /**
   * Perform a HTTP GET request.
   * @param  {String} path
   * @param  {Object} params
   * @return {Promise}
   */
  get (path, params = {}) {
    return this.sendApiRequest('get', path, { params })
  }

  /**
   * Perform a HTTP POST request.
   * @param  {String} path
   * @param  {Object} data
   * @return {Promise}
   */
  post (path, data = {}) {
    return this.sendApiRequest('post', path, data)
  }

  /**
   * Perform a HTTP PUT request.
   * @param  {String} path
   * @param  {Object} data
   * @return {Promise}
   */
  put (path, data = {}) {
    return this.sendApiRequest('put', path, data)
  }

  /**
   * Perform a HTTP PATCH request.
   * @param  {String} path
   * @param  {Object} data
   * @return {Promise}
   */
  patch (path, data = {}) {
    return this.sendApiRequest('patch', path, data)
  }

  /**
   * Perform a HTTP DELETE request.
   * @param  {String} path
   * @param  {Object} params
   * @return {Promise}
   */
  delete (path, params = {}) {
    return this.sendApiRequest('delete', path, { params })
  }

  /**
   * Performs a request to the API using the headers that are expected by the network.
   * @param  {String} method
   * @param  {String} path
   * @param  {Object} payload
   * @return {Promise}
   * @throws Will throw an error if the HTTP request fails.
   */
  sendApiRequest (method, path, payload) {
    return this.sendRequest(method, `/api/${path}`, payload)
  }

  /**
   * Performs a request using the headers that are expected by the network.
   * @param  {String} method
   * @param  {String} path
   * @param  {Object} payload
   * @return {Promise}
   * @throws Will throw an error if the HTTP request fails.
   */
  sendRequest (method, path, payload) {
    this.headers['API-Version'] = this.version

    const client = axios.create({
      baseURL: this.host,
      headers: this.headers,
      timeout: this.timeout
    })

    return client[method](path, payload)
  }
}


examples/send-to-random-wallet.js
#!/usr/bin/env node

const Client = require('@arkecosystem/client')
const crypto = require('@arkecosystem/crypto')

const sendGift = async () => {
  const version = 2
  const amount = 1 * Math.pow(10, 8) // 1 DPHANTOM

  let client

  /**
   * Return a random member of an Array
   * @param {Array} - array
   * @return 1 item of the array
   */
  const getRandom = array => {
    const index = Math.floor(Math.random() * array.length)
    return array[index]
  }

  /**
   * Print an error and exit the app
   * @param {String} - message
   * @param {(Object|Error)} - error
   */
  const fail = (message, error = {}) => {
    // API error message or Error instance message
    const originalError = error.error ? error.error : error.message
    console.error(message, originalError)
    process.exit(1)
  }

  /**
   * Get all the information of a wallet by its address
   * @param {String} address
   * @return {Object} wallet with all the available properties
   */
  const getWallet = async address => {
    try {
      const { data } = await client.resource('wallets').get(address)
      return data.data
    } catch (error) {
      fail(`An error has occured while retrieving the wallet with address ${address}:`, error)
    }
  }

  /**
   * Get a random wallet
   * @return {Object} wallet, but only with some properties
   */
  const getRandomWallet = async () => {
    try {
      const { data } = await client.resource('wallets').all()
      const wallets = data.data
      return getRandom(wallets)
    } catch (error) {
      fail('An error has occured while retrieving the wallets:', error)
    }
  }

  /**
   * Build a transaction using the crypto library and send it with the client
   * @param {String} recipient - the address of the recipient
   * @param {String} senderPublicKey
   * @param {String} passphrase - the sender passphrase
   * @param {Object} the transaction that has been posted
   */
  const postTransaction = async ({ recipient, senderPublicKey, passphrase }) => {
    try {
      const transaction = crypto.transactionBuilder.transfer()
        .amount(amount)
        .vendorField('This is a lovely gift!')
        .recipientId(recipient)
        .senderPublicKey(senderPublicKey)
        .sign(passphrase)
        .getStruct()

      client.resource('transactions').create({ transactions: [transaction] })

      return transaction
    } catch (error) {
      fail('An error has occured while posting the transaction:', error)
    }
  }

  try {
    client = await Client.connect('devnet', version)
  } catch (error) {
    fail('An error has occured while trying to connect to the network:', error.message)
  }

  const sender = process.env.PHANTOM_CLIENT_EXAMPLE_SENDER
  if (!sender) {
    fail('It is necessary to establish the value of the environment variable "PHANTOM_CLIENT_EXAMPLE_SENDER" to an address')
  }

  const passphrase = process.env.PHANTOM_CLIENT_EXAMPLE_PASS
  if (!sender) {
    fail('It is necessary to establish the value of the environment variable "PHANTOM_CLIENT_EXAMPLE_PASS" to the passphrase of the "PHANTOM_CLIENT_EXAMPLE_SENDER" address')
  }

  const senderWallet = await getWallet(sender)

  if (senderWallet.balance < amount) {
    fail(`The sender wallet (${senderWallet.address}) does not have enough balance (${senderWallet.balance} DPHANTOM)`)
  }

  const recipientWallet = await getRandomWallet()
  const recipient = recipientWallet.address

  const transaction = await postTransaction({ recipient, senderPublicKey: senderWallet.address, passphrase })
  const explorerUrl = 'https://dexplorer.phantom.org/transaction/'
  console.log(`\n\tThe gift has been sent to ${recipient} (${explorerUrl}${transaction.id})\n`)
}

sendGift()


lib/index.js
const shuffle = require('lodash.shuffle')
const cloneDeep = require('lodash.clonedeep')
const HttpClient = require('./http')
const resources = require('./resources')
const initialPeers = require('./peers')
const sortPeers = require('./utils/sort-peers')

const mergePeerArrays = (oldArray, newArray) => {
  const peers = []
  for (const oldPeer of oldArray) {
    const newPeer = newArray.find(peer => (peer.ip === oldPeer.ip)) || {}
    peers.push({ ...oldPeer, ...newPeer })
  }
  for (const newPeer of newArray) {
    if (!peers.find(peer => (peer.ip === newPeer.ip))) {
      peers.push(newPeer)
    }
  }

  return peers
}

const getPeerVersion = (peer) => {
  return !peer.version || /^1\./.test(peer.version) ? 1 : 2
}

module.exports = class ApiClient {
  /**
   * Finds all the available peers, sorted by block height and delay
   * @param {String}    network - Network name ('devnet' or 'mainnet')
   * @param {Number}    [version=2] - API version
   * @param {Object[]}  [peersOverride] - List of peers to use instead of initialPeers
   * @return {Object[]}
   */
  static async findPeers (network, version = 2, peersOverride) {
    if (peersOverride === undefined && !initialPeers.hasOwnProperty(network)) {
      throw new Error(`Network "${network}" is not supported`)
    }

    const networkPeers = peersOverride || initialPeers[network]

    // Shuffle the peers to avoid connecting always to the first ones
    shuffle(networkPeers)

    const selfIps = ['127.0.0.1', '::ffff:127.0.0.1', '::1']
    const versionStatus = ['OK', 200]

    // Connect to each peer to get an updated list of peers until a success response
    const client = new ApiClient('http://', version)
    client.http.setTimeout(5000)

    let checks = 0
    let peers = []
    for (const peer of networkPeers) {
      const scheme = peer.isHttps ? 'https://' : 'http://'
      const peerUrl = `${scheme}${peer.ip}:${peer.port}`

      // This method should not crash when a peer fails
      try {
        client.http.host = peerUrl
        // On V2, this endpoint returns the port used by `core-p2p``
        const { data } = await client.resource('peers').all()

        let responsePeers = []
        if (version === 2 && data.data.length) {
          responsePeers = data.data
        }

        if (responsePeers.length) {
          // Ignore local and unavailable peers
          responsePeers = responsePeers.filter(peer => {
            return selfIps.indexOf(peer.ip) === -1 && versionStatus.includes(peer.status) && getPeerVersion(peer) === version
          })

          peers = mergePeerArrays(peers, responsePeers)

          checks++
          if (checks >= 2) {
            break
          }
        }
      } catch (error) {
        // TODO only if a new feature to enable logging is added
        // console.log(`Cannot find find peers of \`${peerUrl}\``)
      }
    }

    // Return at least the initial (hardcoded) peers
    return cloneDeep(sortPeers(peers && peers.length ? peers : networkPeers))
  }

  /**
   * Only for V2
   * Get the configuration of a peer
   * @param {String} host - URL of the host (using `core-p2p` port)
   * @return {(Object|null)}
   */
  static async fetchPeerConfig (host) {
    const httpClient = new HttpClient(host, 2)

    try {
      const { data } = await httpClient.sendRequest('get', '/config')
      if (data) {
        return data.data
      }
    } catch (error) {
      // TODO only if a new feature to enable logging is added
      // console.log(`Error on \`${host}\``)
    }

    return null
  }

  /**
   * Only for V2
   * Check each peer until finding one that has `core-api` enabled and return it,
   * but with the port that is configured to `core-api`
   * @param {Object[]} peers - List of peers to check
   * @return {(Object|null)}
   */
  static async selectApiPeer (peers) {
    for (const peer of peers) {
      const host = `http://${peer.ip}:${peer.port}`
      const config = await ApiClient.fetchPeerConfig(host)

      if (config && config.plugins && config.plugins['@phantomchain/core-api'] &&
      config.plugins['@phantomchain/core-api'].enabled) {
        return {
          ...peer,
          port: config.plugins['@phantomchain/core-api'].port
        }
      }
    }

    return null
  }

  /**
   * Connects to a random peer of the network that supports `core-api`
   * @param {String} network - Network name
   * @param {Number} version - API version
   * @param {Object[]} peersOverride - List of peers to use instead of initialPeers
   * @return {ApiClient}
   */
  static async connect (network, version, peersOverride) {
    const peers = await ApiClient.findPeers(network, version, peersOverride)
    let peer = await ApiClient.selectApiPeer(peers)
    if (!peer) {
      throw new Error('No peer with `core-api` enabled has been found')
    }

    return new ApiClient(`http://${peer.ip}:${peer.port}`, version)
  }

  /**
   * @constructor
   * @param {String} host
   * @param {Number} version - API version
   * @return {void}
   */
  constructor (host, version) {
    this.setConnection(host)
    this.setVersion(version || 2)
  }

  /**
   * Create a HTTP connection to the API.
   * @param {String} host
   * @return {void}
   */
  setConnection (host) {
    this.http = new HttpClient(host, this.version)
  }

  /**
   * Get the HTTP connection to the API.
   * @return {(HttpClient|undefined)}
   */
  getConnection () {
    return this.http
  }

  /**
   * Set the API Version.
   * @param {Number}     version
   * @return {ApiClient}
   */
  setVersion (version) {
    if (!version) {
      throw new Error('A valid API version is required')
    }

    this.version = version
    this.http.setVersion(version)

    return this
  }

  /**
   * Create an instance of a version specific resource.
   * @param  {String}   name
   * @return {Resource}
   */
  resource (name) {
    //return new resources[`v${this.version}`][name](this.http)
    return new resources[`api`][name](this.http)
  }
}
